src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -39,6 +39,69 @@ public static string GetUnescapedString(ReadOnlySpan<byte> utf8Source, int idx)
             return utf8String;
         }
 
+        public static bool UnescapeAndCompare(ReadOnlySpan<byte> utf8Source, ReadOnlySpan<byte> other)
+        {
+            byte[] unescapedArray = null;
+
+            Span<byte> utf8Unescaped = utf8Source.Length <= JsonConstants.StackallocThreshold ?
+                stackalloc byte[utf8Source.Length] :
+                (unescapedArray = ArrayPool<byte>.Shared.Rent(utf8Source.Length));
+
+            Unescape(utf8Source, utf8Unescaped, 0, out int written);
+            Debug.Assert(written > 0);
+
+            utf8Unescaped = utf8Unescaped.Slice(0, written);
+            Debug.Assert(!utf8Unescaped.IsEmpty);
+
+            bool result = other.SequenceEqual(utf8Unescaped);
+
+            if (unescapedArray != null)
+            {
+                utf8Unescaped.Clear();
+                ArrayPool<byte>.Shared.Return(unescapedArray);
+            }
+
+            return result;
+        }
+
+        public static bool UnescapeAndCompare(ReadOnlySequence<byte> utf8Source, ReadOnlySpan<byte> other)
+        {
+            byte[] escapedArray = null;
+            byte[] unescapedArray = null;
+
+            int length = checked((int)utf8Source.Length);
```

Comment:
Maybe we don't expect > 2GB sequences, but it seems like if we did, is this overly conservative?  Seems like there would likely be cases where we could process it in pieces.

Reply:
I don't think there is anything to process here since even if the sequence was > 2 GB, the span we are comparing to cannot be. I'll change it to return false if `length > int.maxvalue` before doing the comparison.

Reply:
Actually, nevermind, we can't return false since the source value, once unescaped could shrink by a factor of 6 and hence be rightfully compared/matched by a 2 GB span. Still, I don't think we are being overly conservative here. We still allow a comparison of up to 2 GB. Sure, theoretically (in the worst case), a fully escaped string, up to 12 GB, could match a 2 GB span, but this is quite an edge case.